clippy_lints/src/and_then_then_some.rs

```diff
@@ -0,0 +1,164 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::source::snippet_with_applicability;
+use clippy_utils::{fn_def_id, match_def_path};
+use rustc_errors::Applicability;
+use rustc_hir::def::Res;
+use rustc_hir::{Block, Closure, Expr, ExprKind, FnDecl, HirId, Node, Pat, Path, QPath, Ty};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::declare_lint_pass;
+use rustc_span::Span;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Detects usage of `Option::and_then` and `bool::then_some` that could
+    /// be replaced with `Option::filter`.
+    ///
+    /// ### Why is this bad?
+    /// Needless complexity, uses recent and uncommon stdlib functions instead of
+    /// one older function.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// let x = Some("foo".to_string());
+    /// let _y = x.clone().and_then(|v| v.starts_with('f').then_some(v));
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// let x = Some("foo".to_string());
+    /// let _y = x.clone().filter(|v| v.starts_with('f'));
+    /// ```
+    #[clippy::version = "1.81.0"]
+    pub AND_THEN_THEN_SOME,
+    nursery,
+    "detects usage of and_then and then_some that can be replaced by filter"
+}
+
+// note: `Option::filter` is older than `bool::then_some`,
+// so no msrv check is required.
+declare_lint_pass!(AndThenThenSome => [AND_THEN_THEN_SOME]);
+
+impl<'tcx> LateLintPass<'tcx> for AndThenThenSome {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>) {
+        match expr.kind {
+            ExprKind::MethodCall(_, selfarg, [arg], _) | ExprKind::Call(_, [selfarg, arg]) => {
+                // TODO: check if type of reciever is diagnostic item Option?
+                if is_and_then(cx, expr) {
+                    if let Some((closure_args, predicate)) = then_some_closure_arg(cx, arg) {
+                        show_sugg(cx, expr.span, selfarg, closure_args, predicate);
+                    }
+                }
+            },
+            _ => {},
+        }
+    }
+}
+
+// `|v| X.then_some(v)` -> Some((span"|v|", X))
+fn then_some_closure_arg<'tcx>(cx: &LateContext<'tcx>, expr: &Expr<'tcx>) -> Option<(Span, &'tcx Expr<'tcx>)> {
+    match expr.kind {
+        ExprKind::Closure(Closure {
+            fn_decl: FnDecl {
+                inputs: [Ty { hir_id: arg_id, .. }],
+                ..
+            },
+            body,
+            ..
+        }) => {
+            if let Node::Expr(expr) = cx.tcx.hir_node(body.hir_id) {
+                (peel_closure_body(cx, expr, *arg_id)).map(|body| (cx.tcx.hir().span(*arg_id), body))
+            } else {
+                None
+            }
+        },
+        _ => None,
+    }
+}
+
+fn peel_closure_body<'tcx>(
```

Comment:
Isn't this kinda just a fancier `clippy_utils::peel_blocks`? Is there any reason not to use that?